modules/conversantAnalyticsAdapter.js

```diff
@@ -0,0 +1,564 @@
+    utils.logError(CNVR_CONSTANTS.LOG_PREFIX + "onAdRenderFailed(): Unable to process RENDER FAILED because adId is missing or doesn't match a record in our cache.");
+    return; // Either no adId to match against a bidWon event, or no data saved from a bidWon event that matches the adId
+  }
+  const adIdObj = cnvrHelper.adIdLookup[adId];
+  const adUnitCode = adIdObj['adUnitCode'];
+  const bidderCode = adIdObj['bidderCode'];
+  const auctionId = adIdObj['auctionId'];
+  delete cnvrHelper.adIdLookup[adId]; // cleanup our cache
+
+  if (!bidderCode || !adUnitCode || !auctionId) {
+    utils.logError(CNVR_CONSTANTS.LOG_PREFIX + 'onAdRenderFailed(): Unable to process RENDER FAILED because lookup cache did not have all the data we required.');
+    return;
+  }
+
+  let timestamp = Date.now();
+  if (cnvrHelper.auctionIdTimestampCache[auctionId]) {
+    timestamp = cnvrHelper.auctionIdTimestampCache[auctionId].timeReceived; // Don't delete, could be multiple winners/auction, allow cleanup to handle
+  }
+
+  const renderFailedPayload = cnvrHelper.createPayload('render_failed', auctionId, timestamp);
+  const adUnitPayload = cnvrHelper.createAdUnit();
+  adUnitPayload.bids[bidderCode] = cnvrHelper.initializeBidDefaults();
+  adUnitPayload.bids[bidderCode].eventCodes.push(CNVR_CONSTANTS.RENDER_FAILED);
+  adUnitPayload.bids[bidderCode].message = 'REASON: ' + args.reason + '. MESSAGE: ' + args.message;
+  renderFailedPayload.adUnits[adUnitCode] = adUnitPayload;
+  sendData(renderFailedPayload);
+}
+
+/**
+ * AUCTION END contains bid and no bid info and all of the auction info we need. This sends the bulk of the information
+ * about the auction back to the servers.  It will also check the timeoutCache for any matching bids, if any are found
+ * then they will be removed from the cache and send back with this payload.
+ * @param args AUCTION END payload, fairly large data structure, main objects are 'adUnits[]', 'bidderRequests[]',
+ * 'noBids[]', 'bidsReceived[]'... 'winningBids[]' seems to be always blank.
+ */
+function onAuctionEnd(args) {
+  const auctionId = args.auctionId;
+  if (!auctionId) {
+    utils.logError(CNVR_CONSTANTS.LOG_PREFIX + 'onAuctionEnd(): No auctionId in args supplied so unable to process event.');
+    return;
+  }
+
+  const auctionTimestamp = args.timestamp ? args.timestamp : Date.now();
+  cnvrHelper.auctionIdTimestampCache[auctionId] = { timeReceived: auctionTimestamp };
+
+  const auctionEndPayload = cnvrHelper.createPayload('auction_end', auctionId, auctionTimestamp);
+  // Get bid request information from adUnits
+  if (!Array.isArray(args.adUnits)) {
+    utils.logError(CNVR_CONSTANTS.LOG_PREFIX + 'onAuctionEnd(): adUnits not defined in arguments.');
+    return;
+  }
+
+  args.adUnits.forEach(adUnit => {
+    const cnvrAdUnit = cnvrHelper.createAdUnit();
+    // Initialize bids with bidderCode
+    adUnit.bids.forEach(bid => {
+      cnvrAdUnit.bids[bid.bidder] = cnvrHelper.initializeBidDefaults();
```

Comment:
It is possible that a publisher has more than one bid for a given bidder within an adUnit.

```js
{
   code: 'some-ad-unit-code',
   mediaTypes: {
      banner: {
         sizes: [300,250, 728,90]
      }
   },
   bids: [
      { bidder: 'ix', params: { siteId: 1234, size: [300,250] } },
      { bidder: 'ix', params: { siteId: 1234, size: [728,90] } }
   ]
}

```

I think in this case, this code will always just set it to whatever the last ix bid is in here.

Same for lower in the `nobids` and `bidsrecieved` parts, it will add an event to eventCodes array, but overwrite the `bidPayload` with the last entry in it.

I guess if you do not care and do not want to track such scenario, or just want the last bid in the array to send it's data it is fine.

But maybe something to look into. I know with my experience I see many pubs having several bids in the bids array for the same bidders.